### PR TITLE
Faster creation of symlinks in CMake

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This uses the [file subcommand CREATE_LINK](https://cmake.org/cmake/help/latest/command/file.html#create-link) to generate symbolic links instead of the more expensive EXECUTE_PROCESS. 

**develop**
`time cmake -G Ninja -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`
```
real	0m36.159s
user	0m32.987s
sys	0m26.173s
```
`time cmake .`
```
real	0m30.386s
user	0m21.352s
sys	0m13.764s
```

**file_create_link**
`time cmake -G Ninja -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`
```
real	0m29.924s
user	0m29.164s
sys	0m21.997s
```
`time cmake .`
```
real	0m22.924s
user	0m18.410s
sys	0m9.649s
```

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)
- Build related changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- N/A. Code added or changed in the PR has been clang-formatted
- N/A. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- N/A. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by quantumsteve** [comment]
Using the default make generator:

**develop**
`time cmake -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`
```
real	0m38.012s
user	0m33.724s
sys	0m26.793s
```
`time cmake .`
```
real	0m30.242s
user	0m21.864s
sys	0m13.904s
```

**file_create_link**
`time cmake -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`
```
real	0m30.959s
user	0m30.347s
sys	0m23.133s
```
`time cmake .`
```
real	0m23.883s
user	0m19.180s
sys	0m9.948s
```

File: CMake/macros.cmake [hunk 1]
```diff
@@ -22,6 +22,23 @@ FUNCTION( COPY_DIRECTORY SRC_DIR DST_DIR )
     EXECUTE_PROCESS( COMMAND ${CMAKE_COMMAND} -E copy_directory "${SRC_DIR}" "${DST_DIR}" )
 ENDFUNCTION()
 
+# Create symlinks for a list of files.
+# This uses file(CREATE_LINK ...) when available.
+# Remove function once CMake 3.14 or later is required.
+FUNCTION( SYMLINK_LIST_OF_FILES FILENAMES DST_DIR)
+    IF(${CMAKE_VERSION} VERSION_LESS "3.14")
```
Comment (ye-luo) [inline CMake/macros.cmake:29]: 3.14 was released in May 2019. So our next release will definitely pass this line with our two-year policy. So feel free to raise the minimum requirement and delete all the work-around for release before 3.14.

Comment (ye-luo) [reply CMake/macros.cmake:29]: Actually it was March 2019, already passed 2 year window.

Comment (prckent) [reply CMake/macros.cmake:29]: I also checked this - we were thinking on the same lines.

However if we can readily not break older versions, we should do so. We could add check + warning for <3.14/2 year old versions though, to prompt users to upgrade.

**Review by prckent — APPROVED** [review]
LGTM (note that a non-ORNL approver is required for merge)

Very nice speedup.

File: CMake/macros.cmake (hunk 1 above)
Comment (ye-luo) [reply CMake/macros.cmake:29]: The offload code with Clang also needs 3.14 to make the object target working. I prefer not to spend our time maintaining old cmake workaround which is not covered by our CI or nightly tests anyway. Compilers have longer release cycles and cmake has more frequent release so even more versions to track.

Comment (prckent) [reply CMake/macros.cmake:29]: Object target is a good point. Combined with this link functionality that is a good forcing function. Lets do it: require minimum cmake 3.14.0.

Comment (prckent) [reply CMake/macros.cmake:29]: Merge this and tidy up tomorrow?

**Comment by prckent** [comment]
OK to test

File: CMake/macros.cmake (hunk 1 above)
Comment (ye-luo) [reply CMake/macros.cmake:29]: We can do that.

**Review by ye-luo — APPROVED** [review]

**Pull request merged**